### DISCUSSION TITLE
Multitenancy cache

### DISCUSSION THREAD

**Discussion by dorep** [spatie/laravel-permission#2310] [Q&A] [answered]
Hello,

We have an application, using your laravel-permission package. We also have multitenant environment using this package: https://tenancyforlaravel.com/

We noticed, that we have a permissions cache issue. Permissions cache does not add a tenant id prefix tag, but it just takes a prefix from cache.php config.

We have a cache bootstrapper, which gives a tenant id prefix to the cache keys. Spatie permissions cache does not seem to take this prefix.

Has anyone seen or solved this "issue"? :) is there any work arounds?

Thank you very much!

Best regards,
Dore

**Comment by erikn69** [comment, marked as answer]
Maybe spatie cache key loads before your "cache bootstrapper"
https://github.com/spatie/laravel-permission/blob/384e06f1bdb34717998e7273ed07ae95fc8296cf/src/PermissionServiceProvider.php#L16
https://github.com/spatie/laravel-permission/blob/384e06f1bdb34717998e7273ed07ae95fc8296cf/src/PermissionRegistrar.php#L82

**Reply by dorep** [reply]
Thank you for your answer, this could be the issue, but did not (yet) find a way to set the sort order. So the goal would be to set tenancyserviceprovider before permissionserviceprovider.

**Reply by erikn69** [reply]
Did you try manually changing `PermissionRegistrar::$cacheKey` on your "cache bootstrapper"?
Also #2309 maybe fix that

**Reply by dorep** [reply]
Thanks, that did the trick :)

**Reply by ahgmulder** [reply]
I have the same problem. So what did the trick? where did you make a modification?

**Reply by ahgmulder** [reply]
We don't have a specific cache bootstrapper and are using spatie/multitenancy

**Reply by dorep** [reply]
Hello,
Usually there are some middlewares that execute in some specific order. It would be best to change this key somewhere after tenancy is initiated.

**Reply by ahgmulder** [reply]
What do you mean by changing this key, this is automatically done by the prefix, right?
Or did you change: self::$cacheKey = config('permission.cache.key');

**Reply by ahgmulder** [reply]
IF I change this to:
self::$cacheKey = config('cache.prefix') . '.' . config('permission.cache.key');
it works.... but I changed this in de vendor folder, and is NOT the place to make that change.
What is the right way to make this modification?

**Reply by dorep** [reply]
that is exactly what i did. But i created cache bootstrapper, that did this. (i also didnt want to do it in vendor folder).
Stancl tenancy package has bootstrappers when in need to do such thing when application is booting up. https://tenancyforlaravel.com/docs/v3/tenancy-bootstrappers/ Check if spatie has something similar.

**Reply by erikn69** [reply]
Never modify the files in the vendor, those files change when you do `composer update` and you lose all your modifications

> We don't have a specific cache bootstrapper 

Did you try just making a middleware for that?

**Reply by ahgmulder** [reply]
I know, my temporary solution is change the persmissions config: cache.store to 'array' instead of 'default'.
But this is not what I really want. I don't know how to change this in a Middleware, because when I create one, the PermissionServiceProvider is already loaded, right? Then overrule the cacheKey has no effect?

**Comment by davutkmbr** [comment]
I managed to write a bootstrapper and extend the PermissionRegistrar class.

```php
<?php

declare(strict_types=1);

namespace App\Tenancy\Bootstrappers;

use Illuminate\Contracts\Foundation\Application;
use Spatie\Permission\PermissionRegistrar;
use Stancl\Tenancy\Contracts\TenancyBootstrapper;
use Stancl\Tenancy\Contracts\Tenant;

class PermissionRegistrarBootstrapper implements TenancyBootstrapper
{
    /** @var PermissionRegistrar */
    protected $originalPermissionRegistrar;

    /** @var Application */
    protected $app;

    public function __construct(Application $app)
    {
        $this->app = $app;
    }

    public function bootstrap(Tenant $tenant)
    {
        $this->originalPermissionRegistrar ??= $this->app[PermissionRegistrar::class];

        $this->app->extend(
            abstract: PermissionRegistrar::class,
            closure: fn() => new PermissionRegistrar($this->app['cache']),
        );
    }

    public function revert()
    {
        $this->app->extend(
            abstract: PermissionRegistrar::class,
            closure: fn() => $this->originalPermissionRegistrar,
        );

        $this->originalPermissionRegistrar = null;
    }
}
```

And add the bootstrapper in `config/tenancy.php`

```php
'bootstrappers' => [
    // ...
    Stancl\Tenancy\Bootstrappers\CacheTenancyBootstrapper::class,
    App\Tenancy\Bootstrappers\PermissionRegistrarBootstrapper::class,
],
```

**Reply by adnanhussain9** [reply]
Thanks for sharing. This is the right solution, modifying libraries is never recommended.

**Reply by jasonpaulraj** [reply]
Needed this, thank you!

**Reply by fklee** [reply]
Excellent, exactly what I was looking for, and it work like a charm ! Would keep this concept for another used.

Just thought that the registrar should always called by app().
